### ISSUE TITLE
Revisit wrap around vs. error behavior in unsigned and signed integer sums

### ISSUE THREAD

**Issue by vmarcos** [MaterializeInc/materialize#17758]
It has been [observed](https://github.com/MaterializeInc/materialize/pull/17709#discussion_r1112110469) that there is some inconsistency between the behaviors of wrap around and error in sum aggregations of signed vs. unsigned integers. Additionally, there are inconsistencies between the behaviors under rendering and under constant folding. For example, rendering uses an `i128` accumulator for all signed integer types: https://github.com/MaterializeInc/materialize/blob/d6371fdc5d9adeb960ddc16a31b032926f063f52/src/compute/src/render/reduce.rs#L1220-L1231
Later, the result is subjected to an `as` conversion to `i64` for sums of `smallint` and `int`, while a `numeric` result is returned for sums of `bigint`: https://github.com/MaterializeInc/materialize/blob/d6371fdc5d9adeb960ddc16a31b032926f063f52/src/compute/src/render/reduce.rs#L1376-L1387
The constant-folding behavior, by contrast, is to use a `sum` over `i64`s for the smaller integer types, which potentially wraps around in release builds: https://github.com/MaterializeInc/materialize/blob/d6371fdc5d9adeb960ddc16a31b032926f063f52/src/expr/src/relation/func.rs#L416-L440

This issue is a request to revisit the behavior of sum aggregations for integer types, making it uniform across rendering and constant folding. Additionally, we may wish to ensure that the semantics chosen (error vs. wrap around) matches our compatibility requirements with PostgreSQL.

**Comment by jkosh44** [comment]
I definitely think that unsigned and signed sums should have the same behavior with respect to overflow vs error. However, if the PostgreSQL behavior is to wrap-around, then I'm ok with diverging here and returning an error. The suggestion from the mailing list to use floats instead of singed ints, is very unsatisfying for me.

**Comment by jkosh44** [comment]
PostgreSQL defines the aggregate sum functions here: 
https://github.com/postgres/postgres/blob/8427ce4c379ee3774b8a9aca13ca7f4bfb76ba85/src/include/catalog/pg_aggregate.dat#L53-L67

I'm not entirely sure what all the fields mean, but all the referenced functions can be found in these files

- https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c
- https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/int8.c

**Comment by vmarcos** [comment]
> I definitely think that unsigned and signed sums should have the same behavior with respect to overflow vs error. However, if the PostgreSQL behavior is to wrap-around, then I'm ok with diverging here and returning an error. The suggestion from the mailing list to use floats instead of singed ints, is very unsatisfying for me.

I agree. Here is an illustration of how the behavior is also inconsistent wrt. wrapping vs. error under constant folding vs. dataflow rendering (requires `--unsafe-mode`):
```sql
DROP TABLE IF EXISTS base CASCADE;
DROP VIEW IF EXISTS constant;

CREATE TABLE base (data int, diff bigint);
CREATE MATERIALIZED VIEW data AS SELECT data FROM base, repeat_row(diff);
INSERT INTO base VALUES (1, 9223372036854775807);
INSERT INTO base VALUES (1, 9223372036854775807);
SELECT SUM(data) FROM data;

CREATE VIEW constant AS
SELECT data FROM (
    SELECT * FROM (
        VALUES (1, 9223372036854775807),
               (1, 9223372036854775807)
        ) AS base(data, diff),
        repeat_row(diff)
);
SELECT SUM(data) FROM constant;
```
The above outputs:
```sql
DROP TABLE
DROP VIEW
CREATE TABLE
CREATE MATERIALIZED VIEW
INSERT 0 1
INSERT 0 1
 sum 
-----
  -2
(1 row)

CREATE VIEW
ERROR:  constant folding encountered reduce on collection with non-positive multiplicities
```

**Comment by frankmcsherry** [comment]
> Here is an illustration of how the behavior is also inconsistent wrt. wrapping vs. error under constant folding vs. dataflow rendering (requires `--unsafe-mode`)

Fwiw you can get a version that does not require `--unsafe-mode` by self-`UNION ALL`ing a bunch of CTEs. Not that anyone needs to actually do this, but also no one should feel safe just because the example requires `repeat_rows`.
